src/katsdpmodels/fetch.py

```diff
@@ -0,0 +1,359 @@
+# Copyright (c) 2020, National Research Foundation (SARAO)
+#
+# Licensed under the BSD 3-Clause License (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Fetch models over HTTP."""
+
+import contextlib
+import errno
+import io
+import hashlib
+import re
+import logging
+import os
+import urllib.parse
+from typing import Dict, Tuple, Optional, Type, TypeVar, Any, cast
+from typing_extensions import Protocol
+
+import requests
+import requests_file
+
+from . import models
+
+
+MAX_ALIASES = 30
+_logger = logging.getLogger(__name__)
+_F = TypeVar('_F', bound='Fetcher')
+_M = TypeVar('_M', bound='models.Model')
+
+
+class Session(Protocol):
+    """Generalization of :class:`requests.Session`."""
+
+    def get(self, url: str, **kwargs) -> requests.Response: ...   # pragma: nocover
+    def head(self, url: str, **kwargs) -> requests.Response: ...  # pragma: nocover
+    def close(self) -> None: ...                                  # pragma: nocover
+
+
+class HttpFile(io.RawIOBase):
+    """File-like object that fetches byte ranges via HTTP.
+
+    This requires the server to advertise for support byte-range requests and
+    to provide a Content-Length. It is currently *not* robust against the
+    content changing.
+
+    Raises
+    ------
+    FileNotFoundError
+        HTTP 404 error
+    PermissionError
+        HTTP 403 error
+    OSError
+        Other HTTP errors, or the server doesn't implement the required features.
+    """
+
+    def __init__(self, session: Session, url: str) -> None:
+        self._session = session
+        self._offset = 0
+        # TODO do we need to set Accept-Encoding: none? Not sure how transfer
+        # encoding interact with byte ranges.
+        with session.head(url) as resp:
+            if resp.status_code == 404:
+                raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), url)
+            elif resp.status_code in {401, 403}:
+                raise PermissionError(errno.EACCES, os.strerror(errno.EACCES), url)
+            resp.raise_for_status()
+            if resp.headers.get('Accept-Ranges', 'none') != 'bytes':
+                raise OSError(None, 'Server does not accept byte ranges', url)
+            try:
+                self._length = int(resp.headers['Content-Length'])
+            except (KeyError, ValueError):
+                raise OSError(None, 'Server did not provide Content-Length header', url) from None
```

Comment:
Are we going to be strict about this - there are situations in which Content-Length is not provided by the server, but you can basically try your luck with a range request that you suspect is good. Might be worth allowing this.

Reply:
I would be very surprised to see any resource that supports range requests without having a known Content-Length. And it would probably add some complexities to the implementation because at the moment if the user tries to read past the end of the file we clip the range, but we can't do that without knowing the file size.

Not having a Content-Length in HTTP/1.1 also typically implies some sort of chunked transfer encoding, which I'm not sure is handled at the moment.

Reply:
I'll buy that - what is the rationale for this being an OSError ?